File: core/client-internal/src/main/java/alluxio/client/file/FileSystem.java
/*
 * Licensed to the University of California, Berkeley under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package alluxio.client.file;

import alluxio.AlluxioURI;
import alluxio.Constants;
import alluxio.annotation.PublicApi;
import alluxio.client.ClientContext;
import alluxio.client.file.options.CreateDirectoryOptions;
import alluxio.client.file.options.CreateFileOptions;
import alluxio.client.file.options.DeleteOptions;
import alluxio.client.file.options.ExistsOptions;
import alluxio.client.file.options.FreeOptions;
import alluxio.client.file.options.GetStatusOptions;
import alluxio.client.file.options.ListStatusOptions;
import alluxio.client.file.options.LoadMetadataOptions;
import alluxio.client.file.options.MountOptions;
import alluxio.client.file.options.OpenFileOptions;
import alluxio.client.file.options.RenameOptions;
import alluxio.client.file.options.SetAttributeOptions;
import alluxio.client.file.options.UnmountOptions;
import alluxio.client.lineage.LineageFileSystem;
import alluxio.exception.AlluxioException;
import alluxio.exception.DirectoryNotEmptyException;
import alluxio.exception.FileAlreadyExistsException;
import alluxio.exception.FileDoesNotExistException;
import alluxio.exception.InvalidPathException;

import java.io.IOException;
import java.util.List;

/**
 * Basic file system interface supporting metadata operations and data operations. Developers
 * should not implement this class but extend the default implementation provided by {@link
 * BaseFileSystem} instead. This ensures any new methods added to the interface will be provided
 * by the default implementation.
 */
@PublicApi
public interface FileSystem {

  /**
   * Factory for {@link FileSystem}.
   */
  class Factory {
    public static FileSystem get() {
      if (ClientContext.getConf().getBoolean(Constants.USER_LINEAGE_ENABLED)) {
        return LineageFileSystem.get();
      }
      return BaseFileSystem.get();
    }
  }

  /**
   * Convenience method for {@link #createDirectory(AlluxioURI, CreateDirectoryOptions)} with
   * default options.
   *
   * @param path the path of the directory to create in Alluxio space
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileAlreadyExistsException if there is already a file or directory at the given path
   * @throws InvalidPathException if the path is invalid
   * @throws AlluxioException if an unexpected exception is thrown
   */
  void createDirectory(AlluxioURI path)
      throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException;

  /**
   * Creates a directory.
   *
   * @param path the path of the directory to create in Alluxio space
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileAlreadyExistsException if there is already a file or directory at the given path
   * @throws InvalidPathException if the path is invalid
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void createDirectory(AlluxioURI path, CreateDirectoryOptions options)
      throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #createFile(AlluxioURI, CreateFileOptions)} with default options.
   *
   * @param path the path of the file to create in Alluxio space
   * @return a {@link FileOutStream} which will write data to the newly created file
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileAlreadyExistsException if there is already a file at the given path
   * @throws InvalidPathException if the path is invalid
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  FileOutStream createFile(AlluxioURI path)
      throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException;

  /**
   * Creates a file.
   *
   * @param path the path of the file to create in Alluxio space
   * @param options options to associate with this operation
   * @return a {@link FileOutStream} which will write data to the newly created file
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileAlreadyExistsException if there is already a file at the given path
   * @throws InvalidPathException if the path is invalid
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  FileOutStream createFile(AlluxioURI path, CreateFileOptions options)
      throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #delete(AlluxioURI, DeleteOptions)} with default options.
   *
   * @param path the path to delete in Alluxio space
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws DirectoryNotEmptyException if recursive is false and the path is a nonempty directory
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void delete(AlluxioURI path)
      throws DirectoryNotEmptyException, FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Deletes a file or a directory.
   *
   * @param path the path to delete in Alluxio space
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws DirectoryNotEmptyException if recursive is false and the path is a nonempty directory
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void delete(AlluxioURI path, DeleteOptions options)
      throws DirectoryNotEmptyException, FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #exists(AlluxioURI, ExistsOptions)} with default options.
   *
   * @param path the path in question
   * @return true if the path exists, false otherwise
   * @throws IOException if a non-Alluxio exception occurs
   * @throws InvalidPathException if the path is invalid
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  boolean exists(AlluxioURI path) throws InvalidPathException, IOException, AlluxioException;

  /**
   * Checks whether a path exists in Alluxio space.
   *
   * @param path the path in question
   * @param options options to associate with this operation
   * @return true if the path exists, false otherwise
   * @throws IOException if a non-Alluxio exception occurs
   * @throws InvalidPathException if the path is invalid
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  boolean exists(AlluxioURI path, ExistsOptions options)
      throws InvalidPathException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #free(AlluxioURI, FreeOptions)} with default options.
   *
   * @param path the path to free in Alluxio space
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void free(AlluxioURI path) throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Evicts any data under the given path from Alluxio space, but does not delete the data from the
   * UFS. The metadata will still be present in Alluxio space after this operation.
   *
   * @param path the path to free in Alluxio space
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void free(AlluxioURI path, FreeOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #getStatus(AlluxioURI, GetStatusOptions)} with default options.
   *
   * @param path the path to obtain information about
   * @return the {@link URIStatus} of the file
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  URIStatus getStatus(AlluxioURI path)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Gets the {@link URIStatus} object that represents the metadata of an Alluxio path.
   *
   * @param path the path to obtain information about
   * @param options options to associate with this operation
   * @return the {@link URIStatus} of the file
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  URIStatus getStatus(AlluxioURI path, GetStatusOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #listStatus(AlluxioURI, ListStatusOptions)} with default options.
   *
   * @param path the path to list information about
   * @return a list of {@link URIStatus}s containing information about the files and directories
   *         which are children of the given path
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  List<URIStatus> listStatus(AlluxioURI path)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * If the path is a directory, returns the {@link URIStatus} of all the direct entries in it.
   * Otherwise returns a list with a single {@link URIStatus} element for the file.
   *
   * @param path the path to list information about
   * @param options options to associate with this operation
   * @return a list of {@link URIStatus}s containing information about the files and directories
   *         which are children of the given path
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  List<URIStatus> listStatus(AlluxioURI path, ListStatusOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #loadMetadata(AlluxioURI, LoadMetadataOptions)} with default
   * options.
   *
   * @param path the path for which to load metadata from UFS
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void loadMetadata(AlluxioURI path)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Loads metadata about a path in the UFS to Alluxio. No data will be transferred.
   *
   * @param path the path for which to load metadata from UFS
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given path does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void loadMetadata(AlluxioURI path, LoadMetadataOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #mount(AlluxioURI, AlluxioURI, MountOptions)} with default
   * options.
   *
   * @param alluxioPath an Alluxio path to mount the data to
   * @param ufsPath a UFS path to mount the data from
   * @throws IOException if a non-Alluxio exception occurs
   * @throws AlluxioException if an Alluxio exception occurs
   */
  void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath) throws IOException, AlluxioException;

  /**
   * Mounts a UFS subtree to the given Alluxio path. The Alluxio path is expected not to exist as
   * the method creates it. If the path already exists, a {@link AlluxioException} will be thrown.
   * This method does not transfer any data or metadata from the UFS. It simply establishes the
   * connection between the given Alluxio path and UFS path.
   *
   * @param alluxioPath an Alluxio path to mount the data to
   * @param ufsPath a UFS path to mount the data from
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws AlluxioException if an Alluxio exception occurs
   */
  void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountOptions options)
      throws IOException, AlluxioException;

  /**
   * Convenience method for {@link #openFile(AlluxioURI, OpenFileOptions)} with default options.
   *
   * @param path the file to read from
   * @return a {@link FileInStream} for the given path
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given file does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  FileInStream openFile(AlluxioURI path)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Opens a file for reading.
   *
   * @param path the file to read from
   * @param options options to associate with this operation
   * @return a {@link FileInStream} for the given path
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given file does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  FileInStream openFile(AlluxioURI path, OpenFileOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #rename(AlluxioURI, AlluxioURI, RenameOptions)} with default
   * options.
   *
   * @param src the path of the source, this must already exist
   * @param dst the path of the destination, this path should not exist
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given file does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void rename(AlluxioURI src, AlluxioURI dst)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Renames an existing Alluxio path to another Alluxio path in Alluxio. This operation will be
   * propagated in the underlying storage if the path is persisted.
   *
   * @param src the path of the source, this must already exist
   * @param dst the path of the destination, this path should not exist
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given file does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void rename(AlluxioURI src, AlluxioURI dst, RenameOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #setAttribute(AlluxioURI, SetAttributeOptions)} with default
   * options.
   *
   * @param path the path to set attributes for
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given file does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void setAttribute(AlluxioURI path)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Sets any number of a path's attributes, such as TTL and pin status.
   *
   * @param path the path to set attributes for
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws FileDoesNotExistException if the given file does not exist
   * @throws AlluxioException if an unexpected Alluxio exception is thrown
   */
  void setAttribute(AlluxioURI path, SetAttributeOptions options)
      throws FileDoesNotExistException, IOException, AlluxioException;

  /**
   * Convenience method for {@link #unmount(AlluxioURI, UnmountOptions)} with default options.
   *
   * @param path an Alluxio path, this must be a mount point
   * @throws IOException if a non-Alluxio exception occurs
   * @throws AlluxioException if an Alluxio exception occurs
   */
  void unmount(AlluxioURI path) throws IOException, AlluxioException;

  /**
   * Unmounts a UFS subtree identified by the given Alluxio path. The Alluxio path match a
   * previously mounted path. The contents of the subtree rooted at this path are removed from
   * Alluxio but the corresponding UFS subtree is left untouched.
   *
   * @param path an Alluxio path, this must be a mount point
   * @param options options to associate with this operation
   * @throws IOException if a non-Alluxio exception occurs
   * @throws AlluxioException if an Alluxio exception occurs
   */
  void unmount(AlluxioURI path, UnmountOptions options) throws IOException, AlluxioException;
}

File: cpp/src/FileSystem.cc
/**
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

#include "FileSystem.h"

FileSystem::FileSystem() {
  JniHelper::Start();
  FileSystem::filesystem = JniHelper::CallStaticObjectMethod(
      "alluxio/client/file/FileSystem$Factory", "get",
      "alluxio/client/file/FileSystem");
}

void FileSystem::closeFileSystem() {
  JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
  JniHelper::DeleteObjectRef(FileSystem::filesystem);
}

FileSystem::~FileSystem() {
  FileSystem::closeFileSystem();
}

Status FileSystem::CreateDirectory(const AlluxioURI& path) {
  return FileSystem::callJniBydefaultOption(path, "createDirectory");
}

Status FileSystem::CreateDirectory(const AlluxioURI& path,
                                   const CreateDirectoryOptions& options) {
  return FileSystem::callJniByOption(path, "createDirectory",
                                     options.getOptions());
}

Status FileSystem::CreateFile(const AlluxioURI& path,
                              FileOutStream** outStream) {
  try {
    jobject fileOutStream = JniHelper::CallObjectMethod(FileSystem::filesystem,
        "alluxio/client/file/FileSystem", "createFile",
        "alluxio/client/file/FileOutStream",  path.getObj());
    Status status = JniHelper::AlluxioExceptionCheck();
    if (status.ok()) {
      *outStream = new FileOutStream(fileOutStream);
     }
    return status;
  } catch (std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::CreateFile(const AlluxioURI& path,
                              const CreateFileOptions& options,
                              FileOutStream** outStream) {
  try {
    jobject fileOutStream = JniHelper::CallObjectMethod(FileSystem::filesystem,
        "alluxio/client/file/FileSystem", "createFile",
		"alluxio/client/file/FileOutStream", path.getObj(),
		options.getOptions());
    Status status = JniHelper::AlluxioExceptionCheck();
    if (status.ok()) {
      *outStream = new FileOutStream(fileOutStream);
    }
    return status;
  } catch (std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::DeletePath(const AlluxioURI& path) {
  return callJniBydefaultOption(path, "delete");
}

Status FileSystem::DeletePath(const AlluxioURI& path,
                              const DeleteOptions& options) {
  return callJniByOption(path, "delete", options.getOptions());
}

Status FileSystem::Exists(const AlluxioURI& path, bool* result) {
  try {
    bool res = JniHelper::CallBooleanMethod(FileSystem::filesystem,
                                            "alluxio/client/file/FileSystem",
                                            "exists", path.getObj());
    result = &res;
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::Exists(const AlluxioURI& path, const ExistsOptions& options,
                          bool* result) {
  bool res;
  try {
    res = JniHelper::CallBooleanMethod(FileSystem::filesystem,
                                       "alluxio/client/file/FileSystem",
                                       "exists", path.getObj(),
                                       options.getOptions());
    result = &res;
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::Free(const AlluxioURI& path) {
  return callJniBydefaultOption(path, "free");
}

Status FileSystem::Free(const AlluxioURI& path, const FreeOptions& options) {
  return callJniByOption(path, "free", options.getOptions());
}

Status FileSystem::GetStatus(const AlluxioURI& path, URIStatus** result) {
  try {
    jobject uriStatus = JniHelper::CallObjectMethod(FileSystem::filesystem,
        "alluxio/client/file/FileSystem","getStatus",
        "alluxio/client/file/URIStatus", path.getObj());
    *result = new URIStatus(uriStatus);
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::GetStatus(const AlluxioURI& path,
                             const GetStatusOptions& options,
                             URIStatus** result) {
  try {
    jobject uriStatus = JniHelper::CallObjectMethod(FileSystem::filesystem,
        "alluxio/client/file/FileSystem","getStatus",
        "alluxio/client/file/URIStatus", path.getObj(), options.getOptions());
    *result = new URIStatus(uriStatus);
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::ListStatus(const AlluxioURI& path,
                              std::vector<URIStatus>* result) {
  JniHelper::LocalRefMapType localRefs;
  try {
    jobject uriStatusList =  JniHelper::CallObjectMethod(
        FileSystem::filesystem, "alluxio/client/file/FileSystem", "listStatus",
        "java/util/List", path.getObj());
    localRefs[JniHelper::GetEnv()].push_back(uriStatusList);
    int listSize = JniHelper::CallIntMethod(uriStatusList, "java/util/List",
                                            "size");

    for (int i = 0; i < listSize; i ++) {
      jobject alluxioUriStatus = JniHelper::CallObjectMethod(
          uriStatusList, "java/util/List", "get", "java/lang/Object", i);
      result->push_back(URIStatus(alluxioUriStatus));
    }
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    return Status::jniError(e);
  }
}

Status FileSystem::ListStatus(const AlluxioURI& path,
                              const ListStatusOptions& options,
                              std::vector<URIStatus>* result) {
  try {
    jobject uriStatusList =  JniHelper::CallObjectMethod(
        FileSystem::filesystem, "alluxio/client/file/FileSystem", "listStatus",
        "java/util/List", path.getObj(), options.getOptions());
    localRefs[JniHelper::GetEnv()].push_back(uriStatusList);
    int listSize = JniHelper::CallIntMethod(uriStatusList, "java/util/List",
                                            "size");
    for (int i = 0; i < listSize; i ++) {
      jobject alluxioUriStatus = JniHelper::CallObjectMethod(
          uriStatusList, "java/util/List", "get", "java/lang/Object", i);
      result->push_back(URIStatus(alluxioUriStatus));
    }
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    return Status::jniError(e);
  }
}

Status FileSystem::Mount(const AlluxioURI& alluxioPath,
                         const AlluxioURI& ufsPath) {
  return callJniBydefaultOption(alluxioPath, ufsPath, "mount");
}

Status FileSystem::Mount(const AlluxioURI& alluxioPath,
                         const AlluxioURI& ufsPath,
                         const MountOptions& options) {
  return callJniByOption(alluxioPath, ufsPath, "mount", options.getOptions());
}

Status FileSystem::GetMountTable(
	  std::map<std::string, MountPointInfo>* result) {
  try {
    jobject jMountTable =  JniHelper::CallObjectMethod(FileSystem::filesystem,
        "alluxio/client/file/FileSystem", "getMountTable", "java/util/Map");
    localRefs[JniHelper::GetEnv()].push_back(jMountTable);

    int mapSize = JniHelper::CallIntMethod(jMountTable, "java/util/Map",
                                           "size");
    jobject keySet =  JniHelper::CallObjectMethod(jMountTable, "java/util/Map",
                                                  "keySet", "java/util/Set");
    localRefs[JniHelper::GetEnv()].push_back(keySet);
    jobject keyArray = JniHelper::CallObjectMethod(keySet, "java/util/Set",
                                                   "toArray",
                                                   "[Ljava/lang/Object");
    localRefs[JniHelper::GetEnv()].push_back(keyArray);
    for(int i = 0; i < mapSize; i ++) {
      jobject keyItem = JniHelper::GetEnv()->
          GetObjectArrayElement((jobjectArray) keyArray, i);
      JniHelper::CacheClassName(keyItem, "java/lang/Object");
      std::string key = JniHelper::JstringToString((jstring)keyItem);
      jobject valueItem = JniHelper::CallObjectMethod(jMountTable,
                                                      "java/util/Map", "get",
                                                      "java/lang/Object",
                                                      (jobject)keyItem);
      result->insert(std::make_pair(key, MountPointInfo(valueItem)));
	  JniHelper::DeleteObjectRef(keyItem);
    }
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    return JniHelper::AlluxioExceptionCheck();
  } catch(std::string e) {
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    return Status::jniError(e);
  }
}

Status FileSystem::OpenFile(const AlluxioURI& path, FileInStream** inStream) {
  jobject fileInStream;
  try {
    fileInStream = JniHelper::CallObjectMethod(FileSystem::filesystem,
        "alluxio/client/file/FileSystem","openFile",
        "alluxio/client/file/FileInStream", path.getObj());
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    Status stus = JniHelper::AlluxioExceptionCheck();
    if (stus.ok()) {
      *inStream = new FileInStream(fileInStream);
    }
    return stus;
  } catch (std::string e) {
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
  	return Status::jniError(e);
  }
}

Status FileSystem::OpenFile(const AlluxioURI& path,
                            const OpenFileOptions& options,
                            FileInStream** inStream) {
  jobject fileInStream;
  try {
    fileInStream = JniHelper::CallObjectMethod(
        FileSystem::filesystem, "alluxio/client/file/FileSystem","openFile",
        "alluxio/client/file/FileInStream", path.getObj(), options.getOptions());
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
    Status stus = JniHelper::AlluxioExceptionCheck();
    if (stus.ok()) {
      *inStream = new FileInStream(fileInStream);
    }
    return stus;
  } catch (std::string e) {
    JniHelper::DeleteLocalRefs(JniHelper::GetEnv(), localRefs);
  	return Status::jniError(e);
  }
}

Status FileSystem::Rename(const AlluxioURI& src, const AlluxioURI& dst) {
  return callJniBydefaultOption(src, dst, "rename");
}

Status FileSystem::Rename(const AlluxioURI& src, const AlluxioURI& dst,
                          const RenameOptions& options) {
  return callJniByOption(src, dst, "rename", options.getOptions());
}

Status FileSystem::SetAttribute(const AlluxioURI& path) {
  return callJniBydefaultOption(path, "setAttribute");
}

Status FileSystem::SetAttribute(const AlluxioURI& path,
                                const SetAttributeOptions& options) {
  return callJniByOption(path, "setAttribute", options.getOptions());
}

Status FileSystem::Unmount(const AlluxioURI& path) {
  return callJniBydefaultOption(path, "unmount");
}

Status FileSystem::Unmount(const AlluxioURI& path,
                           const UnmountOptions& options) {
  return callJniByOption(path, "unmount", options.getOptions());
}

Status FileSystem::callJniBydefaultOption(const AlluxioURI& path,
                                          const std::string& methodName) {
  try {    
    JniHelper::CallVoidMethod(FileSystem::filesystem,
                              "alluxio/client/file/FileSystem", methodName,
                              path.getObj());
    return JniHelper::AlluxioExceptionCheck();
  } catch (std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::callJniBydefaultOption(const AlluxioURI& src,
                                          const AlluxioURI& dst,
                                          const std::string& methodName) {
  try {
    JniHelper::CallVoidMethod(FileSystem::filesystem,
                              "alluxio/client/file/FileSystem", methodName,
                              src.getObj(), dst.getObj());
    return JniHelper::AlluxioExceptionCheck();
  } catch (std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::callJniByOption(const AlluxioURI& path,
                                   const std::string& methodName,
                                   const jobject option) {
  try {
    JniHelper::CallVoidMethod(FileSystem::filesystem,
                              "alluxio/client/file/FileSystem", methodName,
                              path.getObj(), option);
    return JniHelper::AlluxioExceptionCheck();
  } catch (std::string e) {
    return Status::jniError(e);
  }
}

Status FileSystem::callJniByOption(const AlluxioURI& src,
                                   const AlluxioURI& dst,
                                   const std::string& methodName,
                                   const jobject option) {
  try {
    JniHelper::CallVoidMethod(FileSystem::filesystem,
                              "alluxio/client/file/FileSystem", methodName,
                              src.getObj(), dst.getObj(), option);
    return JniHelper::AlluxioExceptionCheck();
  } catch (std::string e) {
    return Status::jniError(e);
  }
}
